packages/nextjs/src/app-router/server/auth.ts

```diff
@@ -1,12 +1,80 @@
+import type { SignedInAuthObject, SignedOutAuthObject } from '@clerk/backend';
+import type { CheckAuthorizationWithCustomPermissions } from '@clerk/types';
+import { notFound } from 'next/navigation';
+
 import { authAuthHeaderMissing } from '../../server/errors';
 import { buildClerkProps, createGetAuth } from '../../server/getAuth';
+import type { AuthObjectWithoutResources } from '../../server/types';
 import { buildRequestLike } from './utils';
 
+type AuthSignedIn = AuthObjectWithoutResources<
+  SignedInAuthObject & {
+    /**
+     * @experimental
+     * This function is experimental as it throws a Nextjs notFound error if user is not authenticated or authorized.
+     * In the future we would investigate a way to throw a more appropriate error that clearly describes the not authorized of authenticated status.
+     */
+    protect: (
+      params?:
+        | Parameters<CheckAuthorizationWithCustomPermissions>[0]
+        | ((has: CheckAuthorizationWithCustomPermissions) => boolean),
+    ) => AuthObjectWithoutResources<SignedInAuthObject>;
+  }
+>;
+
+type AuthSignedOut = AuthObjectWithoutResources<
+  SignedOutAuthObject & {
+    /**
+     * @experimental
+     * This function is experimental as it throws a Nextjs notFound error if user is not authenticated or authorized.
+     * In the future we would investigate a way to throw a more appropriate error that clearly describes the not authorized of authenticated status.
+     */
+    protect: never;
+  }
+>;
+
 export const auth = () => {
-  return createGetAuth({
+  const authObject = createGetAuth({
     debugLoggerName: 'auth()',
     noAuthStatusMessage: authAuthHeaderMissing(),
   })(buildRequestLike());
+
+  (authObject as AuthSignedIn).protect = params => {
+    /**
+     * User is not authenticated
+     */
+    if (!authObject.userId) {
+      notFound();
```

Comment:
@BRKalow just double checking - is this what we're doing in all scenarios?
Where did we land with the custom error page discussion? (redirect to same URL with a query param and throw?)